### PR TITLE
Ft6 Class hierarchy proxy loss

### LINKED ISSUE #9: Create a class hierarchy for proxy-loss types: Quadratic/XEntropy/Hinge
## Summary

- Currently we have to do string comparisons multiple times per data row, per iteration, to figure out which proxy loss function to use to compute the gradients.
  - e.g., https://github.com/feedzai/fairgbm/blob/bulk-fairgbm-modifications/src/objective/constrained_recall_objective.hpp#L86
- According to our benchmarks, this can take up around 30% of the total training time.
- We should create a common interface w/ different subclasses to optimize this.
- This has already been started in #18

### PR DESCRIPTION
**Pull request by garsenio**

### CODE REVIEW

File: src/objective/objective_function.cpp [hunk 1]
```diff
@@ -89,7 +102,7 @@ ObjectiveFunction* ObjectiveFunction::CreateObjectiveFunction(const std::string&
   } else if (type == std::string("constrained_cross_entropy")) {
     return new ConstrainedCrossEntropy(strs);
   } else if (type == std::string("constrained_recall_objective")) {
-    return new ConstrainedRecallObjective(strs);
+    return new ConstrainedRecallObjective(strs); //should never be called. only here for consistency with other objective functions
```
Comment (AndreFCruz) [inline src/objective/objective_function.cpp:105]: why was this comment added?

Comment (AndreFCruz) [reply src/objective/objective_function.cpp:105]: I'd say that the FairGBM package itself is separate from its Python API and the Pulse environment. We can't use recall goal on any of those but we can still use it on the C++ module with command line interface or by loading configs from a file.

**Review by AndreFCruz — COMMENTED** [review]
Looks good overall :slightly_smiling_face: 

Minor comments were added

I still want to test it more thoroughly this week while you're in PTO.

**Comment by sgpjesus** [comment]
Good news :) @AlbertoEAF @AndreFCruz @GoncaloArs 

I made a local test with this PR and got the results on the processing times: 
image

We get a great increase in performance (44min cpu time to 11min). In terms of metrics, the results are the same. Just to make sure, I will compare each prediction.

File: src/objective/objective_function.cpp [hunk 2]
```diff
@@ -12,6 +12,10 @@
 #include "constrained_xentropy_objective.hpp"
 #include "constrained_recall_objective.hpp"
 
+#include "proxy-loss-types/constrained_recall_objective_quadratic.hpp"
```
Comment (AndreFCruz) [inline src/objective/objective_function.cpp:15]: Subclasses should encapsulate proxy losses for the constraints, these classes seem to encapsulate the whole recall-objective + some-specific-proxy behavior.

File: src/objective/objective_function.cpp [hunk 3]
```diff
@@ -43,6 +47,15 @@ ObjectiveFunction* ObjectiveFunction::CreateObjectiveFunction(const std::string&
   } else if (type == std::string("constrained_cross_entropy")) {      // Entry-point for FairGBM code!
     return new ConstrainedCrossEntropy(config);
   } else if (type == std::string("constrained_recall_objective")) {   // Entry-point for FairGBM code!
+    if(config.objective_stepwise_proxy == "quadratic") {
+      return new ConstrainedRecallObjectiveQuadratic(config);
```
Comment (AndreFCruz) [inline src/objective/objective_function.cpp:51]: the hierarchy is not supposed to be under the ObjectiveClass; it's a new hierarchy of proxy losses, and those new classes from the hierarchy will be used in both constrained xentropy and constrained recall objectives.

**Review by AndreFCruz — CHANGES_REQUESTED** [review]
There was a bit of a misunderstanding on the class hierarchy structure.

I don't think this current implementation is even called as the standard is using `constrained_cross_entropy` :sweat_smile:

**Comment by AndreFCruz** [comment]
Closed in favor of #35

**Comment by AndreFCruz** [comment]
FPR-proxy-metrics
